File: 0414-third-maximum-number/0414-third-maximum-number.py
class Solution:
  def thirdMax(self, nums: List[int]) -> int:
    a = -math.inf
    b = -math.inf
    c = -math.inf
    for num in nums:
      if num >= a:
        if num > a:
          a,b,c = num,a,b
      elif num >= b:
        if num > b:
          b,c = num,b
      elif num >= c:
          c = num
    return c if math.isfinite(c) else a

File: 0414-third-maximum-number/0414-third-maximum-number.ts
function thirdMax(nums: number[]): number {
  const max: number[] = [];
  for (let num of nums) {
    for (let i = 0; i < 3; i++) {
      if (max[i] === undefined) {
        max[i] = num;
        break;
      } else if (max[i] === num) {
        break;
      } if (max[i] < num) {
        max.splice(i, 0, num);
        if (max.length > 3) max.pop();
        break;
      }
    }
  }
  return max[2] !== undefined ? max[2] : max[0];
};